nncf/torch/search_building_blocks/search_blocks.py

```diff
@@ -0,0 +1,510 @@
+
+def rule_after_removing_block_graph_has_no_hanging_edges(graph: SearchGraph,
+                                                         start_node: SearchGraphNode,
+                                                         end_node: SearchGraphNode):
+    q = deque([start_node])
+    addit_nodes = set()
+    nodes = []
+    current_node = start_node
+    potential_end_nodes = []
+    while len(q) != 0:
+        current_node = q.pop()
+        if current_node.main_id != start_node.main_id:
+            prev_nodes = graph.get_prev_nodes(current_node.node_key)
+            if len(prev_nodes) > 1:
+                for pn in prev_nodes:
+                    if pn.bottom_id < start_node.bottom_id:
+                        return False # there is extra edge
+                    addit_nodes.add(pn)
+        if current_node.node_key == end_node.node_key:
+            continue
+        if current_node.main_id > end_node.main_id:
+            return False
+        if current_node not in nodes:
+            nodes.append(current_node)
+            next_nodes = graph.get_next_nodes(current_node.node_key)
+            if len(next_nodes) == 0:
+                potential_end_nodes.append(current_node)
+            for next_node in next_nodes:
+                q.appendleft(next_node)
+    if len(q) > 0 or len(potential_end_nodes) > 0:
+        return False
+    for node in addit_nodes:
+        if node not in nodes:
+            return False
+    nodes.append(end_node)
+    return True
+
+def rule_not_lead_to_double_edge(sgraph: SearchGraph, start_node: SearchGraphNode, end_node: SearchGraphNode):
+    #pylint: disable=protected-access
+    if start_node.is_dummy:
+        next_end_node = sgraph.get_next_nodes(end_node.node_key)
+        if len(next_end_node) != 0:
+            attr = sgraph._nx_graph.get_edge_data(start_node.node_key, next_end_node[0].node_key)
+        else:
+            attr = None
+    else:
+        pred_start_node = sgraph.get_prev_nodes(start_node.node_key)
+        next_end_node = sgraph.get_next_nodes(end_node.node_key)
+        if len(pred_start_node) != 0 and len(next_end_node) != 0:
+            attr = sgraph._nx_graph.get_edge_data(pred_start_node[0].node_key, next_end_node[0].node_key)
+        else:
+            attr = None
+    return True if attr is None else False
+
+def rule_not_lead_to_duplication_of_activation(sgraph: SearchGraph,
+                                               start_node: SearchGraphNode,
+                                               end_node: SearchGraphNode):
+    pred_start_node = sgraph.get_prev_nodes(start_node.node_key)
+    next_end_node = sgraph.get_next_nodes(end_node.node_key)
+    if len(next_end_node) == 0 or len(pred_start_node) == 0:
+        return True
+    if pred_start_node[0].is_dummy:
+        pred_start_node = sgraph.get_prev_nodes(pred_start_node[0].node_key)
+
+    if 'relu' == pred_start_node[0].node_type[-1] and 'relu' == next_end_node[0].node_type[0]:
```

Comment:
might be dangerous using string , because we usually change how it's mapped

Reply:
I missed it. Thank you!
Fixed.